nixos/gnome-terminal: init

worldofpeace opened the pull request:
This module obsoletes services.gnome3.gnome-terminal-server
as that's a confusing option for users, and sounds internal.
It's much simpler to have a gnome-terminal module.

This module also correctly includes the vte.sh script
required for vte terminals like gnome-terminal to show the
CWD in the window title and preserved across instances.

This is achieved with the options:
* programs.bash.vteIntegration
* programs.zsh.vteIntegration

as it's best to keep this configuration unguarded by gnome3.enable
to support other vte terminals (such as elementary-terminal).
Note that the distinction between Zsh and Bash doesn't include
a different script, as this script only supports those two shells.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
Fixes #40878.

We should probably check if other desktop managers default include a vte terminal and enable
the appropriate option.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jtojnar commented:
There is a lot of non-DE VTE-based terminals, is this needed for them?

Is `gnome-terminal-server` used by something other than GNOME Terminal?

worldofpeace commented:
> Is gnome-terminal-server used by something other than GNOME Terminal?

I considered this, and the fact that gnome terminal's server isn't a separate package entirely makes me believe this is unlikely and that we shouldn't support it.

worldofpeace commented:
> There is a lot of non-DE VTE-based terminals, is this needed for them?

Yes I believe that any terminal that wants https://github.com/NixOS/nixpkgs/issues/40878 feature (cwd in title and preserved across instances) needs this. Do we need documentation?

jtojnar commented:
> Yes I believe that any terminal that wants #40878 feature (cwd in title and preserved across instances) needs this.

I would suggest to split it to a VTE module to make it easier to find.

> Do we need documentation?

Yeah, for the VTE stuff we probably should write something. Maybe I could add it to the “Creating a custom Freedesktop-like DE” chapter I am meaning to write for a manual.

worldofpeace commented:
>  would suggest to split it to a VTE module to make it easier to find.

I didn't really like that idea too strongly, as vte is just a terminal widget for GTK.
So it would be a super "for this" module which I'm not fond of.

See https://github.com/NixOS/nixpkgs/issues/40878#issuecomment-491730881
It'd be
* `programs.vte.enable*Integration` (not a program)
* `vte.enable*Intergration`

I found `environment.gnome3.enable*Integration` to appear more pleasant, but if you have suggestions I'd be happy to hear them.

worldofpeace commented:
I guess `environment.vte.enable*Integration` is the best I can think of.

Infinisil commented on nixos/modules/programs/gnome-terminal.nix:
```diff
@@ -0,0 +1,72 @@
+    })
+
+    (mkIf config.programs.bash.vteIntegration {
+      programs.bash.interactiveShellInit = vteInitSnippet;
```
How about just a single `programs.shell.vteIntegration` which then does
```nix
mkIf vteIntegration {
  programs.bash.interactiveShellInit = vteInitSnippet;
  programs.zsh.interactiveShellInit = vteInitSnippet;
}
```

The user already controls whether bash or zsh is enabled with `programs.{bash,zsh}.enable`

worldofpeace replied:
> The user already controls whether bash or zsh is enabled with programs.{bash,zsh}.enable

Right, so even if `zsh` isn't enabled  `programs.zsh.interactiveShellInit` doesn't do anything unneeded.

However I feel like this would be slightly dissonant with the `programs.{bash,zsh}` interfaces, we could eliminate one extra option but I don't think `programs.shell.*` would have any more use.

worldofpeace committed:
nixos/vte: init

This module correctly includes the vte.sh script
required for vte terminals like gnome-terminal to show the
CWD in the window title and preserved across instances.

This is achieved with the options:
* programs.bash.vteIntegration
* programs.zsh.vteIntegration

as it's best to keep this configuration unguarded by gnome3.enable
to support other vte terminals (such as elementary-terminal).
Note the distinction between Zsh and Bash doesn't include
a different script, as this script only supports those two shells.

worldofpeace committed:
nixos/gnome-terminal: init

This module obsoletes services.gnome3.gnome-terminal-server
as that's a confusing option for users, and sounds internal.
It's much simpler to have a gnome-terminal module.

worldofpeace commented:
@jtojnar Split them up, the vte stuff is under the `config/` directory.

jtojnar commented on nixos/modules/config/vte.nix:
```diff
@@ -0,0 +1,52 @@
+  vteInitSnippet = ''
+    # Show current working directory in VTE terminals window title.
+    # Supports both bash and zsh, requires interactive shell.
+    . ${pkgs.gnome3.vte}/etc/profile.d/vte.sh
```
drop alias

```suggestion
    . ${pkgs.vte}/etc/profile.d/vte.sh
```

jtojnar commented on nixos/modules/programs/gnome-terminal.nix:
```diff
@@ -0,0 +1,42 @@
+
+  cfg = config.programs.gnome-terminal;
+
+  vteInitSnippet = ''
```
Drop this.

worldofpeace commented:
Thanks @jtojnar, going to fix things up for other DE's (mostly pantheon) with the vte module.
